File: apps/api/services/provisioning.py
"""Organization provisioning: default roles, app entitlements, member assignments.

Formalizes the platform's provisioning flow (see ``core/org_permissions.py``
for terminology):

1. When an organization is created — by business registration, the admin API,
   or a product-app onboarding flow — it is provisioned with its default role
   set and an active subscription to the Enterprise directory app.
2. When a member joins — creation, invite accept, or SSO — their membership is
   linked to the org role matching their role name, and they are assigned to
   the Enterprise app (plus the source app when they arrived through one,
   e.g. Couriers registration/invites).

Members added through the Enterprise directory itself get NO product-app
assignments — those are granted explicitly (``apps:assign``).
"""

from __future__ import annotations

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from core.id import generate_id
from core.logging import get_logger
from db.models import Membership, Organization, OrganizationRole, User
from db.repositories.app_assignments import AppAssignmentRepository
from db.repositories.apps import AppRepository
from db.repositories.org_contacts import OrgContactRepository
from db.repositories.org_roles import OrganizationRoleRepository
from db.repositories.prices import PriceRepository
from db.repositories.subscriptions import SubscriptionRepository
from services.billing_customer_sync import enqueue_customer_ensure_for_organization

logger = get_logger(__name__)

# Every org is entitled to the Enterprise directory app; membership in the org
# is what admits a user to it (assignments are still written for consistency).
ENTERPRISE_APP_SLUG = "876-enterprise"

# Billing is the organization's financial plane — invoices, payment methods and
# the customer registry all hang off it — so an org has it from the moment it
# exists, the same way a Google account reaches Drive without a separate
# sign-up. Heavier surfaces stay behind explicit setup.
BILLING_APP_SLUG = "876-billing"

# Provisioned for every new organization regardless of where it signed up.
DEFAULT_ORG_APP_SLUGS: tuple[str, ...] = (ENTERPRISE_APP_SLUG, BILLING_APP_SLUG)


async def provision_org_apps(db: AsyncSession, org_id: str, *, source_app_id: str | None = None) -> list[str]:
    """Subscribe an org to its default apps plus the app it signed up through.

    Every organization gets Enterprise (where it manages itself) and Billing
    (its financial plane) without asking, and additionally the product app the
    signup came from — a courier company that registers through Couriers should
    land in Couriers, not be told it has no subscription.

    Idempotent, and never fails provisioning: a missing app row means a
    partially seeded environment, which is worth shouting about but is not a
    reason to fail somebody's signup.

    Returns the app ids actually subscribed.
    """
    apps = AppRepository(db)
    subscriptions = SubscriptionRepository(db)
    prices = PriceRepository(db)

    app_ids: list[str] = []
    for slug in DEFAULT_ORG_APP_SLUGS:
        app = await apps.get_by_slug(slug)
        if app is None:
            logger.error("provisioning.default_app_missing", org_id=org_id, slug=slug)
            continue
        app_ids.append(app.id)

    # The source app is identified by the API key the request authenticated
    # with, so it cannot be spoofed by a client claiming to be another app.
    if source_app_id is not None and source_app_id not in app_ids:
        app_ids.append(source_app_id)

    provisioned: list[str] = []
    for app_id in app_ids:
        if await subscriptions.get(org_id, app_id) is not None:
            continue
        default_price = await prices.get_default_for_app(app_id)
        await subscriptions.provision(org_id, app_id, default_price.id if default_price else None)
        provisioned.append(app_id)

    if provisioned:
        logger.info("provisioning.org_apps", org_id=org_id, app_ids=provisioned)
    return provisioned


async def provision_organization(
    db: AsyncSession, org_id: str, now: int, *, source_app_id: str | None = None
) -> dict[str, OrganizationRole]:
    """Idempotently provision an org: default roles + app entitlements.

    Returns the org's system roles keyed by name so callers can link the
    creator's membership without a second query.
    """
    roles = await OrganizationRoleRepository(db).seed_defaults(org_id, now)

    await provision_org_apps(db, org_id, source_app_id=source_app_id)

    organization = (await db.scalars(select(Organization).where(Organization.id == org_id))).first()
    if organization is not None:
        await enqueue_customer_ensure_for_organization(db, organization, now)

    return roles


async def ensure_default_contact(db: AsyncSession, org_id: str, user: User, now: int) -> None:
    """Seed the org's default primary contact from its owner.

    Idempotent: does nothing once the org has any active contact. Contacts can
    later be re-pointed, demoted, or extended with non-member people — this
    only guarantees a new org starts with its owner as the primary contact.
    """
    repo = OrgContactRepository(db)
    if await repo.list_by_org(org_id):
        return

    await repo.create(
        id=generate_id("orgContact"),
        organization_id=org_id,
        user_id=user.id,
        first_name=user.first_name,
        last_name=user.last_name,
        type="general",
        is_primary=True,
        email=user.email,
        phone=user.phone,
        created_at=now,
        updated_at=now,
    )
    logger.info("provisioning.default_contact", org_id=org_id, user_id=user.id)


async def resolve_member_permissions(db: AsyncSession, membership: Membership) -> set[str]:
    """Effective org permissions for a membership.

    Prefers the linked organization role; memberships not yet linked (or whose
    role row was removed) fall back to the code-default permission set for
    their role name.
    """
    from core.org_permissions import default_permissions_for_role_name

    if membership.role_id:
        role = await OrganizationRoleRepository(db).get_by_id_for_org(
            membership.role_id, membership.organization_id
        )
        if role is not None:
            return set(role.permissions)
    return set(default_permissions_for_role_name(membership.role))


async def resolve_role_id(db: AsyncSession, org_id: str, role_name: str) -> str | None:
    """The org-role row ID for a role name; None when the org has no such role."""
    role = await OrganizationRoleRepository(db).get_by_name(org_id, role_name)
    return role.id if role else None


async def link_membership_role(db: AsyncSession, membership: Membership, now: int) -> None:
    """Point ``membership.role_id`` at the org role matching its role name."""
    role_id = await resolve_role_id(db, membership.organization_id, membership.role)
    if membership.role_id != role_id:
        membership.role_id = role_id
        membership.updated_at = now
        await db.flush()


async def assign_member_apps(
    db: AsyncSession,
    *,
    org_id: str,
    user_id: str,
    now: int,
    source_app_id: str | None = None,
    assigned_by: str | None = None,
) -> None:
    """Assign a member the Enterprise directory app plus the source app, if any."""
    assignments = AppAssignmentRepository(db)

    enterprise_app = await AppRepository(db).get_by_slug(ENTERPRISE_APP_SLUG)
    if enterprise_app is not None:
        await assignments.assign(org_id, user_id, enterprise_app.id, now, assigned_by=assigned_by)

    if source_app_id and (enterprise_app is None or source_app_id != enterprise_app.id):
        await assignments.assign(org_id, user_id, source_app_id, now, assigned_by=assigned_by)

File: apps/api/src/services/provisioning.ts
import { generateId } from '@/platform/ids'
import { getLogger } from '@/platform/logger'
import {
  DEFAULT_ORG_ROLES,
  defaultPermissionsForRoleName,
} from '@/platform/permissions'
import { dispatchBillingCustomerSyncOnce } from '@/workers/billing-customer-dispatch'

import * as repository from './provisioning.repository'
import type { OrgRoleRow } from './provisioning.repository'
import { enqueueCustomerEnsureForOrganization } from './billing-customer-sync'
import { createBillingCustomerSyncRepository } from './billing-customer-sync.repository'
import {
  enabledProvisioningApplicationSlugs,
  requirePersistedProvisioningPolicy,
  resolveFreshProvisioningPolicy,
  retrievePersistedProvisioningPolicy,
} from './provisioning-policy'

/**
 * Organization provisioning: default roles, database-owned setup entitlements,
 * member assignments, and shared financial infrastructure.
 *
 * Fresh organizations are routed exactly once before their first entitlement
 * write. Existing organizations without a persisted setup are deliberately left
 * on the legacy Enterprise-only behavior until the explicit Phase 2 backfill
 * assigns them a setup.
 */

const log = getLogger('provisioning')

/** Every organization retains the Enterprise directory access plane. */
export const ENTERPRISE_APP_SLUG = '876-enterprise'

/** Standalone Billing product access remains independent of shared finance. */
export const BILLING_APP_SLUG = '876-billing'

/** Legacy-safe entitlement set for organizations not yet explicitly backfilled. */
export const DEFAULT_ORG_APP_SLUGS = [ENTERPRISE_APP_SLUG] as const

/** Apps that currently consume the shared Work service. */
export const WORK_DEPENDENT_APP_SLUGS = ['876-crm'] as const

export type EnqueueCustomerEnsure = (
  organizationId: string,
  now: number
) => Promise<void>

const defaultEnqueueCustomerEnsure: EnqueueCustomerEnsure = async (
  organizationId,
  now
) => {
  try {
    const organization =
      await repository.findOrganizationForCustomerEnsure(organizationId)
    if (!organization) return
    await enqueueCustomerEnsureForOrganization(
      { repository: createBillingCustomerSyncRepository() },
      organization,
      now
    )
    await dispatchBillingCustomerSyncOnce({ limit: 1 })
  } catch (error) {
    log.error(
      { err: error, org_id: organizationId },
      'provisioning.customer_ensure_failed'
    )
  }
}

export async function seedDefaultRoles(
  organizationId: string,
  now: number
): Promise<Record<string, OrgRoleRow>> {
  const existing = new Map(
    (await repository.listRolesForOrg(organizationId)).map((role) => [
      role.name,
      role,
    ])
  )

  const seeded: Record<string, OrgRoleRow> = {}
  for (const definition of DEFAULT_ORG_ROLES) {
    const current = existing.get(definition.name)
    if (current) {
      seeded[definition.name] = current
      continue
    }

    seeded[definition.name] = await repository.createRole({
      id: generateId('role'),
      organizationId,
      name: definition.name,
      displayName: definition.displayName,
      description: definition.description ?? null,
      permissions: [...definition.permissions],
      isSystem: true,
      createdAt: BigInt(now),
      updatedAt: BigInt(now),
    })
  }

  return seeded
}

async function provisionedApplicationSlugs(
  organizationId: string,
  creationTimestamp?: number
): Promise<string[]> {
  let persisted = await retrievePersistedProvisioningPolicy(organizationId)

  if (!persisted && creationTimestamp !== undefined) {
    const fresh = await resolveFreshProvisioningPolicy(
      organizationId,
      creationTimestamp
    )
    if (fresh) persisted = fresh
  }

  return persisted
    ? enabledProvisioningApplicationSlugs(persisted.policy)
    : [...DEFAULT_ORG_APP_SLUGS]
}

/**
 * Subscribe an organization to setup-enabled applications plus an explicit
 * source app. Billing/Invoice are never inferred from finance infrastructure.
 */
export async function ensureOrgAppSubscriptions(
  organizationId: string,
  options: {
    sourceAppId?: string | null
    selectionTimestamp?: number
  } = {}
): Promise<{ appIds: string[]; provisioned: string[] }> {
  const appIds: string[] = []

  for (const slug of await provisionedApplicationSlugs(
    organizationId,
    options.selectionTimestamp
  )) {
    const app = await repository.findAppBySlug(slug)
    if (!app) {
      log.error(
        { org_id: organizationId, slug },
        'provisioning.entitled_app_missing'
      )
      continue
    }
    if (!appIds.includes(app.id)) appIds.push(app.id)
  }

  // The source app comes from the authenticated API-key principal. It is an
  // explicit product request, not a location-policy inference.
  const sourceAppId = options.sourceAppId ?? null
  if (sourceAppId !== null && !appIds.includes(sourceAppId))
    appIds.push(sourceAppId)

  const provisioned: string[] = []
  for (const appId of appIds) {
    const existing = await repository.findSubscription(organizationId, appId)
    if (existing) {
      if (existing.status === 'active' && !existing.hasItems) {
        const defaultPrice = await repository.findDefaultPriceForApp(appId)
        if (defaultPrice)
          await repository.ensureSubscriptionDefaultPrice({
            subscriptionId: existing.id,
            itemId: generateId('subscriptionItem'),
            priceId: defaultPrice.id,
            now: BigInt(Math.floor(Date.now() / 1000)),
          })
      }
      continue
    }

    const defaultPrice = await repository.findDefaultPriceForApp(appId)
    await repository.provisionSubscription({
      id: generateId('subscription'),
      itemId: generateId('subscriptionItem'),
      organizationId,
      appId,
      priceId: defaultPrice?.id ?? null,
      status: 'active',
      now: BigInt(Math.floor(Date.now() / 1000)),
    })
    provisioned.push(appId)
  }

  if (provisioned.length > 0) {
    log.info(
      { org_id: organizationId, app_ids: provisioned },
      'provisioning.org_apps'
    )
  }

  return { appIds, provisioned }
}

export async function ensureOrgAppsFinanceReady(
  organizationId: string,
  options: { appIds?: string[] } = {}
): Promise<void> {
  const [{ ensureAppReady }, { createFinanceProvisioningRepository }] =
    await Promise.all([
      import('./finance-provisioning-readiness'),
      import('./finance-provisioning.repository'),
    ])

  const appIds =
    options.appIds ?? (await repository.listSubscribedAppIds(organizationId))
  const deps = { repository: createFinanceProvisioningRepository() }

  for (const appId of appIds) {
    await ensureAppReady(deps, { organizationId, appId })
  }
}

export async function provisionOrgApps(
  organizationId: string,
  options: { sourceAppId?: string | null; now?: number } = {}
): Promise<string[]> {
  const { appIds, provisioned } = await ensureOrgAppSubscriptions(
    organizationId,
    {
      sourceAppId: options.sourceAppId ?? null,
      selectionTimestamp: options.now,
    }
  )
  await ensureOrgAppsFinanceReady(organizationId, { appIds })
  return provisioned
}

/**
 * Idempotently provision roles, setup app entitlements, customer-registry
 * identity, and finance readiness.
 *
 * A fresh organization is selected automatically when its `createdAt` equals
 * the bootstrap timestamp passed here. Older unselected organizations are not
 * routed implicitly; callers that require Phase 2 semantics use
 * `requireProvisioningSelection` and receive a stable 409 until backfill.
 */
export async function provisionOrganization(
  organizationId: string,
  now: number,
  options: {
    sourceAppId?: string | null
    enqueueCustomerEnsure?: EnqueueCustomerEnsure
    deferFinanceReadiness?: boolean
    requireProvisioningSelection?: boolean
  } = {}
): Promise<Record<string, OrgRoleRow>> {
  if (options.requireProvisioningSelection) {
    await requirePersistedProvisioningPolicy(organizationId)
  } else {
    await resolveFreshProvisioningPolicy(organizationId, now)
  }

  const roles = await seedDefaultRoles(organizationId, now)
  const { appIds } = await ensureOrgAppSubscriptions(organizationId, {
    sourceAppId: options.sourceAppId ?? null,
    selectionTimestamp: now,
  })

  const organization = await repository.findOrganization(organizationId)
  if (organization)
    await (options.enqueueCustomerEnsure ?? defaultEnqueueCustomerEnsure)(
      organizationId,
      now
    )

  if (!options.deferFinanceReadiness)
    await ensureOrgAppsFinanceReady(organizationId, { appIds })

  return roles
}

export async function ensureDefaultContact(
  organizationId: string,
  user: {
    id: string
    firstName: string
    lastName: string | null
    email: string | null
    phone: string | null
  },
  now: number
): Promise<void> {
  if ((await repository.listOrgContacts(organizationId)).length > 0) return

  await repository.createOrgContact({
    id: generateId('orgContact'),
    organizationId,
    userId: user.id,
    firstName: user.firstName,
    lastName: user.lastName,
    type: 'general',
    isPrimary: true,
    email: user.email,
    phone: user.phone,
    createdAt: BigInt(now),
    updatedAt: BigInt(now),
  })

  log.info(
    { org_id: organizationId, user_id: user.id },
    'provisioning.default_contact'
  )
}

export async function resolveMemberPermissions(membership: {
  roleId: string | null
  organizationId: string
  role: string
}): Promise<Set<string>> {
  if (membership.roleId) {
    const role = await repository.findRoleForOrg(
      membership.roleId,
      membership.organizationId
    )
    if (role) return new Set(role.permissions)
  }

  return new Set(defaultPermissionsForRoleName(membership.role))
}

export async function resolveRoleId(
  organizationId: string,
  roleName: string
): Promise<string | null> {
  const role = await repository.findRoleByName(organizationId, roleName)
  return role?.id ?? null
}

export async function linkMembershipRole(
  membership: {
    id: string
    organizationId: string
    role: string
    roleId: string | null
  },
  now: number
): Promise<void> {
  const roleId = await resolveRoleId(membership.organizationId, membership.role)
  if (membership.roleId === roleId) return

  await repository.updateMembershipRole(membership.id, roleId, BigInt(now))
}

/**
 * Assign the member every setup-enabled application plus the explicit source
 * app. Existing assignments are upserted, so retries are safe.
 */
export async function assignMemberApps(params: {
  organizationId: string
  userId: string
  now: number
  sourceAppId?: string | null
  assignedBy?: string | null
}): Promise<void> {
  const now = BigInt(params.now)
  const assignedBy = params.assignedBy ?? null
  const appIds: string[] = []

  for (const slug of await provisionedApplicationSlugs(
    params.organizationId,
    params.now
  )) {
    const app = await repository.findAppBySlug(slug)
    if (app && !appIds.includes(app.id)) appIds.push(app.id)
  }

  const sourceAppId = params.sourceAppId ?? null
  if (sourceAppId && !appIds.includes(sourceAppId)) appIds.push(sourceAppId)

  for (const appId of appIds) {
    await repository.assignApp({
      id: generateId('appAssignment'),
      organizationId: params.organizationId,
      userId: params.userId,
      appId,
      assignedBy,
      now,
    })
  }
}
